drivers: gsm adding support for rssi readout for the PPP link

mtahirbutt opened the pull request:
This PR addresses radio signal stength measurement during and before
    PPP session. The PR provides provides facility of readout for both +CSQ and +CESQ
    versions depending upon the modems.
    This PR follows the idea of rssi readout of PR#35496.
    Additionally, this PR also considers the issue #33773 and it has been observed with 
    SARA R410M modem, the avoiding sending AT+COPS=0,0 improves the
    performance slightly if the modem is already in automatic operator selection mode.

hwilmers commented on drivers/modem/gsm_ppp.c:
```diff
@@ -474,6 +611,32 @@ static void gsm_finalize_connection(struct gsm_modem *gsm)
 	}
+
+#if defined(CONFIG_GSM_MUX)
+		k_work_schedule(&rssi_work_handle, K_NO_WAIT);
```
This should only be called if gsm muxing actually is enabled. To ensure this, I suggest to call it right after modem_cmd_handler_tx_unlock().

mtahirbutt replied:
addressed now

mtahirbutt committed:
drivers: modem: RSSI measurement for PPP link

This PR addresses radio signal stength measurement during
and before PPP session. The PR provides provides facility
of readout for both +CSQ and +CESQ versions depending
upon the modems. This PR follows the idea of rssi readout
of PR#35496. Additionally, reliable Cell info update
is also ensured.

Signed-off-by: Tahir Akram <mtahirbutt@hotmail.com>

hwilmers commented on drivers/modem/gsm_ppp.c:
```diff
@@ -492,12 +630,58 @@ static void gsm_finalize_connection(struct gsm_modem *gsm)
+		while (counter < GSM_CELL_INFO_MAX_RETRIES && ret < 0) {
+			ret = gsm_query_cellinfo(gsm);
+			counter++;
+		}
```
Another possibility would be to read cell info in rssi_handler(). Since cell info can change during the connection, this would be an advantage. The ublox driver is implementing it like this, and by calling the handler once during initialisation it also avoids some code duplication.

mtahirbutt replied:
addressed now, I have considered both MUX and without MUX possibilities.

jukkar commented on drivers/modem/Kconfig.gsm:
```diff
@@ -78,6 +78,20 @@ config MODEM_GSM_MANUAL_MCCMNO
 	  the network provider and may need to be changed if auto is not
 	  selected.
 
+config MODEM_GSM_RSSI_WORK_PERIOD
```
Better would be call this `MODEM_GSM_RSSI_POLLING_PERIOD` or similar, as that immediately tells reader what the variable is controlling.

mtahirbutt replied:
thanks for the review. addressed now

jukkar commented on drivers/modem/Kconfig.gsm:
```diff
@@ -78,6 +78,20 @@ config MODEM_GSM_MANUAL_MCCMNO
 	  selected.
 
+config MODEM_GSM_RSSI_WORK_PERIOD
+	int "Configure RSSI WORK polling period"
```
Please add some info about the time resolution.
So something like `Configure RSSI polling period (in seconds)`,
also the "WORK" does not need to be mentioned here as it does not give any useful info.

mtahirbutt replied:
thanks, addressed now

jukkar commented on drivers/modem/Kconfig.gsm:
```diff
@@ -78,6 +78,20 @@ config MODEM_GSM_MANUAL_MCCMNO
+
+config MODEM_GSM_ENABLE_CESQ_RSSI
+	bool "Enable +CESQ RSSI measurement"
+	default n
```
Options are not set by default so this can be removed.

mtahirbutt replied:
addressed, thanks

jukkar commented on drivers/modem/Kconfig.gsm:
```diff
@@ -78,6 +78,19 @@ config MODEM_GSM_MANUAL_MCCMNO
 	  selected.
 
+config MODEM_GSM_RSSI_POLLING_PERIOD
+	int "Configure RSSI polling period"
```
Please add info about the resolution either here in the prompt, or in the help text. Otherwise it is impossible for the casual reader to understand what the value means. I would recommend adding `(in seconds)` to the prompt.

jukkar reviewed (approved):
LGTM
